docs/rfcs/031-sharding-static.md

```diff
@@ -0,0 +1,407 @@
+when we implement shard splitting in future, it will be useful to enable shards to reference layers
+written by other shards (specifically the parent shard during a split), so that shards don't
+have to exhaustively copy all data into their own shard-prefixed keys.
+
+To enable this, the `IndexPart` structure will be extended to store the (shard number, shard count)
+tuple on each layer, such that it can construct paths for layers written by other shards. This
+naturally raises the question of who "owns" such layers written by ancestral shards: this problem
+will be addressed in phase 2.
+
+For backward compatibility, any index entry without shard information will be assumed to be
+in the legacy shardidentity.
+
+#### WAL Ingest
+
+In Phase 1, all shards will subscribe to the safekeeper to download WAL content. They will filter
+it down to the pages relevant to their shard:
+
+- For ordinary user data writes, only retain a write if it matches the ShardIdentity
+- For metadata describing relations etc, all shards retain these writes.
+
+The pageservers must somehow give the safekeeper correct feedback on remote_consistent_lsn:
+one solution here is for the 0th shard to periodically peek at the IndexParts for all the other shards,
+and have only the 0th shard populate remote_consistent_lsn. However, this is relatively
+expensive: if the safekeeper can be made shard-aware then it could be taught to use
+the max() of all shards' remote_consistent_lsns to decide when to trim the WAL.
+
+#### Compaction/GC
+
+No changes needed.
+
+The pageserver doesn't have to do anything special during compaction
+or GC. It is implicitly operating on the subset of keys that map to its ShardIdentity.
+This will result in sparse layer files, containing keys only in the stripes that this
+shard owns. Where optimizations currently exist in compaction for spotting "gaps" in
+the key range, these should be updated to ignore gaps that are due to sharding, to
+avoid spuriously splitting up layers ito stripe-sized pieces.
+
+### Compute Endpoints
+
+Compute endpoints will need to:
+
+- Accept a vector of connection strings as part of their configuration from the control plane
+- Route pageserver requests according to mapping the hash of key to the correct
+  entry in the vector of connection strings.
+
+Doing this in compute rather than routing requests via a single pageserver is
+necessary to enable sharding tenants without adding latency from extra hops.
+
+### Control Plane
+
+Tenants, or _Projects_ in the control plane, will each own a set of TenantShards (this will
+be 1 for small tenants). Logic for placement of tenant shards is just the same as the current logic for placing
+tenants.
+
+Tenant lifecycle operations like deletion will require fanning-out to all the shards
+in the tenant. The same goes for timeline creation and deletion: a timeline should
+not be considered created until it has been created in all shards.
```

Comment:
How pageserver HTTP API will change? Are we going to rename `tenant_id` to `tenant_shard_id` there or it seems easier to add shard field in places where it is appropriate? How to create tenant with multiple shards now? Should we mention shard index and total shard count on creation?

Reply:
> Are we going to rename tenant_id to tenant_shard_id there or it seems easier to add shard field in places where it is appropriate? 

Because TenantShardId is backward & forward compatible with TenantId in its encoding, we can just substitute one for the other in existing APIs.  For example all the timeline APIs already accept a TenantShardId in the path.

> How to create tenant with multiple shards now?

That can be done via the sharding service, but I don't expect the control plane to do that.  Because we're skipping static sharding and moving straight to shard splitting for GA, the control plane can just create tenants with 1 shard, and the sharding service will take care of splitting them when they grow.

Reply:
But sharding service will also do it through some http API, right? Is there a doc for it?

Reply:
It's `TenantCreateRequest::shard_parameters`, only used in tests.  In prod I expect to use shard splitting, so that tenants will mostly start with just 1 shard, and auto-grow.